### PR TITLE
Improve ExecutableWhiskAction coverage and general typesafety

### PR DESCRIPTION
**Pull request by markusthoemmes**
- Improves the implementation of ExecutableWhiskAction to conform closer to existing implementations.
- Use ExecutableWhiskAction as early as possible to statically check only the right types reach the Invoker.

### CODE REVIEW

File: common/scala/src/main/scala/whisk/core/entity/WhiskAction.scala [hunk 1]
```diff
@@ -71,6 +71,32 @@ case class WhiskActionPut(
     }
 }
 
+abstract class WhiskActionLike(override val name: EntityName) extends WhiskEntity(name) {
```
Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/WhiskAction.scala]: why did you need to introduce this intermediate type?

File: common/scala/src/main/scala/whisk/core/entity/WhiskAction.scala [hunk 2]
```diff
@@ -163,25 +163,30 @@ case class WhiskAction(
...
-    name: EntityName,
+    override val name: EntityName,
     exec: CodeExec[_],
+    parameters: Parameters = Parameters(),
     limits: ActionLimits = ActionLimits(),
     version: SemVer = SemVer(),
-    rev: DocRevision = DocRevision()) {
+    publish: Boolean = false,
+    annotations: Parameters = Parameters())
+    extends WhiskActionLike(name) {
```
Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/WhiskAction.scala]: i see. :/

Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/entity/WhiskAction.scala]: I felt uncomfortable with the difference between `WhiskAction` and `ExecutableWhiskAction` so I opted for a slightly more verbose but satisfying approach.

**Comment by rabbah** [comment]
I'm deferring this until some related PRs that refactor related code are resolved. I can address resultant conflicts at that point.
